### PR TITLE
Update CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by rrzhang139**
### Description

Updating contribution guidelines for more information.

### Issue Link

N/A

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- Best Coding Practices Section instructing streamlined guidelines how to Open PRs, submit issues
- Adding Tests
- FAQs

### Rationale

More thorough documentation on how to contribute, as well as more resources/standards on how to contribute. Previous guidelines are not as streamlined or structured as this one. 

### How Has This Been Tested?

N/A

### CODE REVIEW

**Comment by rrzhang139** [comment]
@ChihChengLiang thoughts on new contributor guide?

File: CONTRIBUTING.md
Comment (ChihChengLiang) [inline CONTRIBUTING.md:30]: If we assume the contributor needs this detailed git command guide, then squashing commits might be difficult for them. We can make the commits squashing optional.

**Review by ChihChengLiang — COMMENTED** [review]
Sorry, I thought the PR was at the drafting stage.
I left a comment on squashing.

**Review by ChihChengLiang — APPROVED** [review]
LGTM.

**Comment by ChihChengLiang** [comment]
Would be great if @ed255 or @aguzmant103 can give it a look.

File: CONTRIBUTING.md
Comment (ed255) [inline CONTRIBUTING.md:15]: I would rephrase this a bit: We encourage everyone to open an issue before getting to work on a PR.  This will be a must for "PRs that change significantly the codebase or the way on which some procedure ..."

File: CONTRIBUTING.md
Comment (ed255) [inline CONTRIBUTING.md:16]: Following the steps chronologically, I think this point should go after current point 7.

File: CONTRIBUTING.md
Comment (ed255) [reply CONTRIBUTING.md:30]: We must also consider that currently for outdated branches in a PR, github shows a button called "Update branch" that merges main into the branch (instead of doing the rebase).  This seems to be very convenient for reviewers who want to merge a PR from a contributor without write access; because after an approval they can just click, wait for tests, and merge.  Maybe we could rephrase points 8 and 9 as suggestions?

File: CONTRIBUTING.md
Comment (ed255) [inline CONTRIBUTING.md:38]:
```suggestion
        git rebase main
        git push -f
```
I think these steps are simpler and don't require an interactive rebase

File: CONTRIBUTING.md
Comment (ed255) [inline CONTRIBUTING.md:30]:
```suggestion
9. Please do not merge main into your branch as you develop your pull request; instead, rebase your branch on top of the latest main if your pull request branch is long-lived.
    1. How to regularly update your PR with main:
```

File: CONTRIBUTING.md
Comment (ed255) [inline CONTRIBUTING.md:44]: Which forums?  Is this a reference to https://github.com/privacy-scaling-explorations/zkevm-circuits/discussions ?  If so, could you add a link?

File: CONTRIBUTING.md
Comment (ed255) [inline CONTRIBUTING.md:46]: Does this apply in our case? zkevm-circuits is more a library than a program, so usually there are no commands.

File: CONTRIBUTING.md
Comment (ed255) [inline CONTRIBUTING.md:70]: > Also, if any changes are requested, bring back the PR to `Draft` until it's ready again with the aforementioned requirements.

We're not usually doing this.  I think having unresolved requests from a reviewer, or having the PR marked as changes requested is enough to signal that the PR is not ready to be merged and the PR submitter has work to do.

File: CONTRIBUTING.md
Comment (ed255) [inline CONTRIBUTING.md:92]: I tried following the discord link and it tells me "Invite Invalid"

**Review by ed255 — APPROVED** [review]
LGTM! Thanks for addressing the requests :)

**Pull request merged**
